### PR TITLE
Support mover dry run

### LINKED ISSUE #3574: cluster move --dry-run ?
**User Story**

I want to test whether I can move something from bootstrap to target cluster
is it reasonable to provide a comment to test what to be moved and what will be kept but
not really move them? 

if it's reasonable, I can work on it

/kind feature

### PR DESCRIPTION
**Pull request by jichenjc**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3574 

**Release note**:
```release-note
ObjectMover interface function has been updated to `Move(namespace string, toCluster Client, dryRun bool) error`
from `Move(namespace string, toCluster Client) error`
```

### CODE REVIEW

File: cmd/clusterctl/client/cluster/mover.go [hunk 1]
```diff
@@ -200,8 +211,10 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
 
 	// Ensure all the expected target namespaces are in place before creating objects.
 	log.V(1).Info("Creating target namespaces, if missing")
-	if err := o.ensureNamespaces(graph, toProxy); err != nil {
-		return err
+	if o.isDryrun {
```
Comment (JoelSpeed) [inline cmd/clusterctl/client/cluster/mover.go:214]: Surely we only want to do this if this isn't dry run? Think this might be a mistake

**Review by JoelSpeed — COMMENTED** [review]
What sort of output do you get when you do a move with dry run? Does it list all of the resources that it would have moved or something?

What's the use case for this option, would be good to have this noted in the PR description. Is there a related issue or discussion about this feature somewhere?

File: cmd/clusterctl/client/cluster/mover.go (hunk 1 above)
Comment (jichenjc) [reply cmd/clusterctl/client/cluster/mover.go:214]: @JoelSpeed I am still working on it to clean up some stuffs , I updated the PR title with the issue and 
yes, the purpose is to only list resource to be moved but not actually take any actions..

File: cmd/clusterctl/client/cluster/client.go [hunk 2]
```diff
@@ -102,6 +102,7 @@ type clusterClient struct {
 	repositoryClientFactory RepositoryClientFactory
 	pollImmediateWaiter     PollImmediateWaiter
 	processor               yaml.Processor
+	isDryrun                bool
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/client.go:105]: I'm a little debated about having `isDryrun` as a field in the `clusterClient`, because this is an option that applies only to the `Move` command.
Is there a particular reason for this approach? or should we consider to scope down the change to the single command?

File: cmd/clusterctl/client/cluster/mover.go [hunk 3]
```diff
@@ -43,6 +43,7 @@ type ObjectMover interface {
 type objectMover struct {
 	fromProxy             Proxy
 	fromProviderInventory InventoryClient
+	isDryrun              bool
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/mover.go:46]: Same as above only with s/clusterClient/objectMover (should we consider to scope down the change to the single command?)

File: cmd/clusterctl/client/cluster/mover.go [hunk 4]
```diff
@@ -76,8 +83,10 @@ func (o *objectMover) Move(namespace string, toCluster Client) error {
 	// This is required because if the infrastructure is provisioned, then we can reasonably assume that the objects we are moving are
 	// not currently waiting for long-running reconciliation loops, and so we can safely rely on the pause field on the Cluster object
 	// for blocking any further object reconciliation on the source objects.
-	if err := o.checkProvisioningCompleted(objectGraph); err != nil {
-		return err
+	if !o.isDryrun {
+		if err := o.checkProvisioningCompleted(objectGraph); err != nil {
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/mover.go:87]: Being a dry-run, I think we should avoid also `checkTargetProviders`. WDYT?

Adding some more context, I'm wondering if we still need to access to a target cluster when dry-running or we should support dry running without providing a `--to-kubeconfig` argument

File: cmd/clusterctl/client/cluster/mover.go [hunk 5]
```diff
@@ -200,8 +210,10 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
 
 	// Ensure all the expected target namespaces are in place before creating objects.
 	log.V(1).Info("Creating target namespaces, if missing")
-	if err := o.ensureNamespaces(graph, toProxy); err != nil {
-		return err
+	if !o.isDryrun {
+		if err := o.ensureNamespaces(graph, toProxy); err != nil {
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/mover.go:214]: When dry running, I think that we should also avoid to pause/unpause the clusters, wdyt?

**Review by fabriziopandini — COMMENTED** [review]
@jichenjc thanks for this PR
I have left some comments, mostly related to having dryrun as a ClusterClient field vs adding a parameter to the Move operation.
On top of that two notes
- Have you considered if we can make it possible/easy to implement the backup use case, which is related to this PR? (see https://github.com/kubernetes-sigs/cluster-api/issues/3574#issuecomment-685452626)
- I would like to have a test verifying that dry runs actually does not move objects

**Comment by jichenjc** [comment]
@fabriziopandini thanks for the review, I actually tried to add a param in 
https://github.com/kubernetes-sigs/cluster-api/blob/master/cmd/clusterctl/client/cluster/mover.go#L39

to `Move(namespace string, toCluster Client, dryrun bool) error`

then functional it works ,but `scripts/ci-apidiff.sh` complains the API change is breaking but I don't know why it's breaking so I have to add that param in struct directly to avoid the error above

**Comment by jichenjc** [comment]
@fabriziopandini thanks for the additional comments.. so let's wait for additional comments and see how we can move forward, in addition, I need learn how to fix api-diff issue as well (to make the test pass)

**Comment by wfernandes** [comment]
>@jichenjc the apidiff job is optional, and as of today for clusterctl as a library we are not providing a compatibility guarantee (see https://github.com/kubernetes-sigs/cluster-api/blob/master/cmd/clusterctl/README.md)
> Personally I think that changing the move signature in the cluster package is fine, because we are going to preserve compatibility on the CLI for move and on the Move operation (1 additional flag, no breaking changes) in the client package (1 additional field to MoveOptions, no breaking changes)

Agreed to what @fabriziopandini said. However, we should probably document the change as part of the release notes so that users are aware. 

@jichenjc I haven't yet tested this locally on my machine but I'd love to see some tests asserting the dry run has the expected behavior.

**Comment by wfernandes** [comment]
@jichenjc We'll also need to update documentation for this.

**Comment by jichenjc** [comment]
@wfernandes works for me , I will do what you mentioned above and I assume only need update releaes notes  ,and no document needed, correct?

**Comment by wfernandes** [comment]
@jichenjc I was thinking it would be more visible to users if we added it to our book (https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html). 
Updating this page would require an update here: https://github.com/kubernetes-sigs/cluster-api/blob/feaff31a93174f79929f1264f47f8ad057577454/docs/book/src/clusterctl/commands/move.md

But yes, including in our release notes would be helpful as well. 🙂

**Comment by jichenjc** [comment]
@wfernandes got it , appreciate the info~

File: cmd/clusterctl/client/cluster/mover.go (hunk 5 above)
Comment (jichenjc) [reply cmd/clusterctl/client/cluster/mover.go:214]: yes, will add it

File: cmd/clusterctl/client/cluster/mover.go (hunk 4 above)
Comment (jichenjc) [reply cmd/clusterctl/client/cluster/mover.go:87]: um.. this need further check, I will consider it

**Comment by fabriziopandini** [comment]
sgtm for me.
I would like to have a test verifying that dry runs actually does not move objects, after then IMO we can remove WIP

**Comment by jichenjc** [comment]
@fabriziopandini  I don't know how to fix the api-diff issue, looks to me it's comparing with some code ...can you provide some guide?

**Comment by fabriziopandini** [comment]
@jichenjc api-diff is not blocking, just informative.
In the case of this PR, we agreed on the changes to the `ObjectMover.Move` signature. As suggested by @wfernandes we should probably document the change as part of the release notes.

**Comment by jichenjc** [comment]
@fabriziopandini I am live with that and will update release NOTEs
the question I am having is after the merge, the apidiff will always fail? then if we break the test why we need run test from then on ? if @vincepri @wfernandes any comments?

**Comment by fabriziopandini** [comment]
> the question I am having is after the merge, the apidiff will always fail? 

No, apidiff compares commits in a PR to the tip of the main branch. Once a PR is merged, the new is considered as a new baseline for following PRs.

**Comment by jichenjc** [comment]
@fabriziopandini thanks, Now I understand it 
I will do additional test here and remove the WIP soon

File: cmd/clusterctl/client/move.go [hunk 6]
```diff
@@ -63,7 +66,7 @@ func (c *clusterctlClient) Move(options MoveOptions) error {
 		options.Namespace = currentNamespace
 	}
 
-	if err := fromCluster.ObjectMover().Move(options.Namespace, toCluster); err != nil {
+	if err := fromCluster.ObjectMover().Move(options.Namespace, toCluster, options.DryRun); err != nil {
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/move.go:69]: In case of dry-run, is it possible to avoid to create the toCluster/runningEnsureCRSs on it?

File: cmd/clusterctl/cmd/move.go [hunk 7]
```diff
@@ -78,6 +81,7 @@ func runMove() error {
 		FromKubeconfig: client.Kubeconfig{Path: mo.fromKubeconfig, Context: mo.fromKubeconfigContext},
 		ToKubeconfig:   client.Kubeconfig{Path: mo.toKubeconfig, Context: mo.toKubeconfigContext},
 		Namespace:      mo.namespace,
+		DryRun:         mo.dryRun,
```
Comment (fabriziopandini) [inline cmd/clusterctl/cmd/move.go:84]: In case of dry run, is it possible to tolerate the `toKubeconfig` to be empty?

Comment (jichenjc) [reply cmd/clusterctl/cmd/move.go:84]: I need more check here, good suggestion

File: cmd/clusterctl/client/move.go (hunk 6 above)
Comment (jichenjc) [reply cmd/clusterctl/client/move.go:69]: yes, let me think about it

Comment (jichenjc) [reply cmd/clusterctl/client/move.go:76]: updated just now

File: cmd/clusterctl/cmd/move.go (hunk 7 above)
Comment (jichenjc) [reply cmd/clusterctl/cmd/move.go:85]: updated as well, seems works on my env

```
# ../clusterctl move  -v 10  --dry-run
No default config file available
Performing move...
********************************************************
This is a dryrun mover, will not perform any real action
********************************************************

```

File: cmd/clusterctl/client/cluster/mover_test.go [hunk 8]
```diff
@@ -452,6 +452,81 @@ func Test_getMoveSequence(t *testing.T) {
...
+
+			err = mover.move(graph, toProxy)
+			if tt.wantErr {
+				g.Expect(err).To(HaveOccurred())
+				return
+			}
+
+			g.Expect(err).NotTo(HaveOccurred())
+
+			// check that the objects are removed from the source cluster and are created in the target cluster
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/mover_test.go:488]: This comment does not reflect the logic below

File: cmd/clusterctl/client/cluster/mover_test.go [hunk 9]
```diff
@@ -452,6 +452,81 @@ func Test_getMoveSequence(t *testing.T) {
...
+				if err == nil {
+					if oFrom.GetNamespace() != "" {
+						t.Errorf("%v created in target cluster which should not", key)
+						continue
+					}
+				} else if !apierrors.IsNotFound(err) {
+					t.Errorf("error = %v when checking for %v should not created ojects in target cluster", err, key)
+					continue
+				}
+
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/mover_test.go:524]:
```suggestion
```

**Review by fabriziopandini — APPROVED** [review]
/approve
Apart for a small nit, IMO you can now remove hold from this PR
@wfernandes, please take another look to this PR

File: cmd/clusterctl/client/cluster/mover_test.go (hunk 8 above)
Comment (jichenjc) [reply cmd/clusterctl/client/cluster/mover_test.go]: updated

**Comment by jichenjc** [comment]
@fabriziopandini thanks :) , updated

**Comment by jichenjc** [comment]
@wfernandes @vincepri can you help to check this as well? Thanks

**Review by wfernandes — COMMENTED** [review]
@jichenjc Thanks for adding the test. The output of the command looks informative. 

I just have some really minor nits.
**TLDR:** In the documentation we should probably alert the user to use log levels required in order to get the dry-run results. WDYT?

No Log level
```
$ clusterctl move --dry-run
Performing move...
********************************************************
This is a dryrun mover, will not perform any real action
********************************************************
Discovering Cluster API objects
Moving Cluster API objects Clusters=1
Creating objects in the target cluster
Deleting objects from the source cluster
```

Log level >5
```
$ clusterctl move --dry-run -v5
Using configuration File="/Users/wfernandes/.cluster-api/clusterctl.yaml"
Performing move...
********************************************************
This is a dryrun mover, will not perform any real action
********************************************************
Discovering Cluster API objects
AWSMachineTemplate Count=2
KubeadmConfigTemplate Count=1
KubeadmControlPlane Count=1
Cluster Count=1
KubeadmConfig Count=2
MachineDeployment Count=1
Machine Count=2
MachineSet Count=1
AWSCluster Count=1
Secret Count=8
AWSMachine Count=2
Total objects Count=22
Excluding secret from move (not linked with any Cluster) name="default-token-4mpt8"
Moving Cluster API objects Clusters=1
Pausing the source cluster
Creating target namespaces, if missing
Creating objects in the target cluster
Creating Cluster="wff-test" Namespace="default"
Creating AWSCluster="wff-test" Namespace="default"
Creating KubeadmConfigTemplate="wff-test-md-0" Namespace="default"
Creating KubeadmControlPlane="wff-test-control-plane" Namespace="default"
Creating MachineDeployment="wff-test-md-0" Namespace="default"
Creating AWSMachineTemplate="wff-test-control-plane" Namespace="default"
Creating AWSMachineTemplate="wff-test-md-0" Namespace="default"
Creating Machine="wff-test-control-plane-s9hsw" Namespace="default"
Creating MachineSet="wff-test-md-0-5955cfb58d" Namespace="default"
Creating Secret="wff-test-ca" Namespace="default"
Creating Secret="wff-test-etcd" Namespace="default"
Creating Secret="wff-test-kubeconfig" Namespace="default"
Creating Secret="wff-test-proxy" Namespace="default"
Creating Secret="wff-test-sa" Namespace="default"
Creating AWSMachine="wff-test-control-plane-qv95s" Namespace="default"
Creating Machine="wff-test-md-0-5955cfb58d-hbk5z" Namespace="default"
Creating KubeadmConfig="wff-test-control-plane-96gj6" Namespace="default"
Creating KubeadmConfig="wff-test-md-0-ggklr" Namespace="default"
Creating Secret="wff-test-control-plane-96gj6" Namespace="default"
Creating AWSMachine="wff-test-md-0-qrc94" Namespace="default"
Creating Secret="wff-test-md-0-ggklr" Namespace="default"
Deleting objects from the source cluster
Deleting Secret="wff-test-md-0-ggklr" Namespace="default"
Deleting KubeadmConfig="wff-test-md-0-ggklr" Namespace="default"
Deleting Secret="wff-test-control-plane-96gj6" Namespace="default"
Deleting AWSMachine="wff-test-md-0-qrc94" Namespace="default"
Deleting AWSMachine="wff-test-control-plane-qv95s" Namespace="default"
Deleting Machine="wff-test-md-0-5955cfb58d-hbk5z" Namespace="default"
Deleting KubeadmConfig="wff-test-control-plane-96gj6" Namespace="default"
Deleting Machine="wff-test-control-plane-s9hsw" Namespace="default"
Deleting MachineSet="wff-test-md-0-5955cfb58d" Namespace="default"
Deleting Secret="wff-test-ca" Namespace="default"
Deleting Secret="wff-test-etcd" Namespace="default"
Deleting Secret="wff-test-kubeconfig" Namespace="default"
Deleting Secret="wff-test-proxy" Namespace="default"
Deleting Secret="wff-test-sa" Namespace="default"
Deleting AWSCluster="wff-test" Namespace="default"
Deleting KubeadmConfigTemplate="wff-test-md-0" Namespace="default"
Deleting KubeadmControlPlane="wff-test-control-plane" Namespace="default"
Deleting MachineDeployment="wff-test-md-0" Namespace="default"
Deleting AWSMachineTemplate="wff-test-control-plane" Namespace="default"
Deleting AWSMachineTemplate="wff-test-md-0" Namespace="default"
Deleting Cluster="wff-test" Namespace="default"
Resuming the target cluster
Using configuration File="/Users/wfernandes/.cluster-api/clusterctl.yaml"
⚠️  Using a development build of clusterctl. CLIVersion="0.3.9-76-9c6dbeecafc799" LatestGithubRelease="v0.3.9"
```
---------

File: docs/book/src/clusterctl/commands/move.md [hunk 10]
```diff
@@ -64,3 +64,7 @@ This can now be achieved with the following procedure:
 
 > Note: It's required to have at least one worker node to schedule Cluster API workloads (i.e. controllers).
 > A cluster with a single control plane node won't be sufficient due to the `NoSchedule` taint. If a worker node isn't available, `clusterctl init` will timeout.
+
+## Dry run
+
+with `--dry-run` option you can dry run the move action without really take any actual actions but only print logs.
```
[the remark this answers is not in the document]
Comment (jichenjc) [reply docs/book/src/clusterctl/commands/move.md:70]: ok, will update

File: cmd/clusterctl/client/cluster/mover_test.go [hunk 11]
```diff
@@ -452,6 +452,80 @@ func Test_getMoveSequence(t *testing.T) {
...
+				oFrom := &unstructured.Unstructured{}
+				oFrom.SetAPIVersion(node.identity.APIVersion)
+				oFrom.SetKind(node.identity.Kind)
+
+				if err := csFrom.Get(ctx, key, oFrom); err != nil {
+					t.Errorf("error = %v when checking for %v kept in source cluster", err, key)
+					continue
+				}
+
+				// objects are not created int target cluster as it's dry run
```
[the remark this answers is not in the document]
Comment (jichenjc) [reply cmd/clusterctl/client/cluster/mover_test.go:509]: done

File: cmd/clusterctl/client/move.go [hunk 12]
```diff
@@ -29,6 +33,9 @@ type MoveOptions struct {
 	// Namespace where the objects describing the workload cluster exists. If unspecified, the current
 	// namespace will be used.
 	Namespace string
+
+	// DryRun means the move action is a dry run, not a real action will be performed
```
[the remark this answers is not in the document]
Comment (jichenjc) [reply cmd/clusterctl/client/move.go:37]: done

File: cmd/clusterctl/client/cluster/mover.go [hunk 13]
```diff
@@ -36,27 +36,36 @@ import (
...
 var _ ObjectMover = &objectMover{}
 
-func (o *objectMover) Move(namespace string, toCluster Client) error {
+func (o *objectMover) Move(namespace string, toCluster Client, dryRun bool) error {
 	log := logf.Log
 	log.Info("Performing move...")
+	o.dryRun = dryRun
+	if o.dryRun {
+		log.Info("********************************************************")
+		log.Info("This is a dryrun mover, will not perform any real action")
```
[the remark this answers is not in the document]
Comment (jichenjc) [reply cmd/clusterctl/client/cluster/mover.go:58]: ok, got it
